### ISSUE TITLE
[C++] Implement ODBC driver "wrapper" using FlightSQL

### ISSUE THREAD

**Issue by asfimport** [apache/arrow#30622]
The ODBC analogue to ARROW-7744

**Reporter**: [Wes McKinney](https://issues.apache.org/jira/browse/ARROW-15111) / @wesm
#### Related issues:
- [[FlightRPC][Java] Implement Flight SQL JDBC Driver](https://github.com/apache/arrow/issues/23982) (relates to)

<sub>**Note**: *This issue was originally created as [ARROW-15111](https://issues.apache.org/jira/browse/ARROW-15111). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

**Comment by spencerwilson** [comment]
Possibly related to dremio/warpdrive:
- https://docs.dremio.com/current/sonar/client-applications/drivers/arrow-flight-sql-odbc-driver
- https://github.com/dremio/flightsql-odbc , which may or may not be up-to-date source code for ^

**Comment by alinaliBQ** [comment]
https://github.com/dremio/flightsql-odbc should have the Apache 2.0 license, my understanding is that it is a data source client that works with Apache Flight SQL. One should be able to develop an ODBC driver using this data source client. 

Would [`cpp/src/arrow/flight/sql`](https://github.com/apache/arrow/tree/main/cpp/src/arrow/flight/sql) be an appropriate subdirectory to put the ODBC driver? James and I are thinking that this looks like the right place, but please let us know if the proposal makes sense. 
cc @jduo

**Comment by alinaliBQ** [comment]
Hello @wesm and @lidavidm, would either of you mind taking a look and letting me know if [cpp/src/arrow/flight/sql](https://github.com/apache/arrow/tree/main/cpp/src/arrow/flight/sql) is an appropriate subdirectory to put the ODBC driver?
Thank you

**Comment by lidavidm** [comment]
@alinaliBQ how about `cpp/src/flightsql_odbc` or similar?

**Comment by alinaliBQ** [comment]
@lidavidm Sure, `cpp/src/flightsql_odbc` makes sense to me. Ty.
I have another question. We are looking to build a new ODBC driver for Flight SQL that can be part of the Arrow project. It would utilize parts of the [Amazon Timestream ODBC driver](https://github.com/awslabs/amazon-timestream-odbc-driver) and the [Flight SQL ODBC driver](https://github.com/dremio/flightsql-odbc) (written by Dremio), which are Open Source and Apache 2.0-licensed. Are there any questions/concerns regarding using those drivers?

**Comment by lidavidm** [comment]
Do you mean that you plan to import or copy large chunks of one or both projects, or do you mean that you plan to use them as dependencies? If the former, I think depending on how much is copied we may have to think about IP clearance, but it's not clear to me what the threshold is.

**Comment by alinaliBQ** [comment]
We're planning the former, so large chunks from both projects will be used. We're in the designing stage of developing the driver, so things may change later; the plan is that `flightsql-odbc` will be mostly used as-is other than changes to conforming to Arrow coding guidelines, and for Amazon Timestream driver, only its ODBC function entry code will be used and adapted to call into `flightsql-odbc` classes.

**Comment by lidavidm** [comment]
Ok. I would encourage you to submit a PR ASAP even if it is not complete so we can do as much development as possible in Apache repos. The guidelines state IP clearance is needed when most of the development is done outside of Apache repos, so submitting a single large PR (as with the JDBC driver) might mean we want to do an IP clearance to be safe; submitting PRs here early and often would help avoid that.

At least, flightsql-odbc is probably enough code that we might want to do IP clearance anyways...however, please discuss this on dev@ so others can chime in

**Comment by alinaliBQ** [comment]
We were thinking of submitting the PR early as well. Our initial plan is to submit the PR when the driver is able to connect the Flight SQL ODBC driver, with irrelevant code pruned. James has let me know that the community has indicated that the PR for Timestream ODBC + flightsql-odbc can be sent even if the driver doesn't compile since it's for starting the IP scanning process, so we can go with that. And I have written an email @dev for other's opinions on this matter at: https://lists.apache.org/thread/t1r3pntpzoxdncgoj5f581hxyyl19bkl.

**Comment by laurentgo** [comment]
> Possibly related to dremio/warpdrive:
> 
>     * https://docs.dremio.com/current/sonar/client-applications/drivers/arrow-flight-sql-odbc-driver
>       
>       * the linked [driver download page](https://www.dremio.com/drivers/odbc/) indicates that its license is some version of LGPL, but I can't find a link to its source code

The driver is a combination of ASL 2.0 and LGPL. The LGPL license is available at https://github.com/dremio/warpdrive/blob/master/license.txt

**Comment by alinaliBQ** [comment]
Hi @lidavidm, currently our team's implementation is being done inside our own [arrow fork](https://github.com/Bit-Quill/arrow/pulls). I was wondering if you know any Arrow community members who would be interested in taking a look at our incremental PRs for the ODBC driver?  If so, we could assign folks as the reviewers to let them know which PRs to take a look at. We would appreciate additional pairs of eyes from the community. Please let me know if you have any questions.

**Comment by kou** [comment]
Could you open a PR to apache/arrow instead of your fork?

**Comment by wesm** [comment]
We could create a branch on apache/arrow so that PRs do not have to go into the main branch (in case things are unstable)?

**Comment by kou** [comment]
It seems that you already have many changes.
Can we break down it to small pieces and proceed step-by-step like we did to implement Google Cloud Storage file system and Azure Blob Storage file system?
* The parent issue for Google Cloud Storage file system: https://issues.apache.org/jira/browse/ARROW-1231
* The parent issue for Azure Blob Storage file system: https://github.com/apache/arrow/issues/18014

**Comment by kou** [comment]
> We could create a branch on apache/arrow so that PRs do not have to go into the main branch (in case things are unstable)?

We can do it but we may not need to do it. Because we will have a build option for this module such as `ARROW_FLIGHT_SQL_ODBC` and it's `OFF` by default. If this module isn't built by default. We don't need to care about stability.

We require IP clearance for this, right? We can't merge the first PR to apache/arrow before it's completed.
I think that we should avoid developing outside of apache/arrow as much as possible. So I think that we should focus on the IP clearance instead of developing for now.

**Comment by lidavidm** [comment]
I agree with Kou. We can consider a branch if needed but since this should be reasonably fenced off from the rest of the codebase, it should be OK to just have it on main.

@alinaliBQ for the original question please tag me to start with and we can pull in more people as needed.

**Comment by devozerov** [comment]
Our team has some experience working with Dremio's ODBC driver when connecting to a custom Arrow Flight endpoint (a Trino fork with Arrow Flight SQL support). We were also considering taking Dremio's Apache 2.0 code as a base and creating a fully-fledged driver. I am very happy that some work already being done in Arrow's community. If needed, we can help with the review.

**Comment by vic-tsang** [comment]
Hi all, 

I am in the process of preparing mac and linux installers for the Apache Arrow Flight SQL ODBC driver. I would like to gather inputs from the community regarding the install path directory. Most ODBC installers install in the following directories 

- /Library 
- /opt
- /usr/local

Here are some potential install directory path suggestions 

- /Library/Apache/ArrowFlightSQLODBC
- /opt/apache/arrowflightsqlodbc

Please let me know what you think. If you have other ideas, please let me know. Thanks in advance!

**Comment by lidavidm** [comment]
@vic-tsang you may want to also send a message to the mailing lists!

**Comment by kou** [comment]
For Linux, how about using deb/rpm instead of installers?

Could you also share directory structure under the prefix? We install only `${PREFIX}/lib/libXXX.so`?

**Comment by alinaliBQ** [comment]
Hi Kou, yup the plan is to use the deb/rpm packages for Linux systems. By `$PREFIX`, are you referring to the `PREFIX` environment variable? Is there any docs that we could refer to in order to better understand the `PREFIX` variable?

**Comment by kou** [comment]
> By `$PREFIX`, are you referring to the `PREFIX` environment variable? Is there any docs that we could refer to in order to better understand the `PREFIX` variable?

No. I just want to refer `/Library`, `/opt`, `/usr/local`, ... in the https://github.com/apache/arrow/issues/30622#issuecomment-3216041435 by `${PREFIX}`.
